flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/RecordWriter.java

```diff
@@ -290,11 +290,11 @@ protected void addBufferConsumer(BufferConsumer consumer, int targetChannel) thr
 	}
 
 	@VisibleForTesting
-	public BufferBuilder getBufferBuilder() throws IOException, InterruptedException {
-		BufferBuilder builder = targetPartition.tryGetBufferBuilder();
+	public BufferBuilder getNewBufferBuilder(int targetChannel) throws IOException, InterruptedException {
```

Comment:
Hmmm, this is getting a bit messy. We have three different methods `getBufferBuilder()`/`getNewBufferBuilder(targetChannel)`, `getBufferBuilder(targetChannel)` and `requestNewBufferBuilder(targetChannel)`. (It's not your fault, I've missed that in a previous code review which introduced this `getBufferBuilder()` method)

I'm not sure how to name those things, but maybe this method should be named `requestNewBufferBuilder(targetChannel)` (it would no longer be an abstract method), and this method should overridden in the two sub classes in the following way:

```
	@Override
	public BufferBuilder ChannelSelectorRecordWriter#requestNewBufferBuilder(int targetChannel) throws IOException, InterruptedException {
		checkState(bufferBuilders[targetChannel] == null || bufferBuilders[targetChannel].isFinished());

		BufferBuilder bufferBuilder = super.requestNewBufferBuilder(targetChannel);
		addBufferConsumer(bufferBuilder.createBufferConsumer(), targetChannel);
		bufferBuilders[targetChannel] = bufferBuilder;
		return bufferBuilder;
	}
```
and
```
	@Override
	public BufferBuilder BroadcastRecordWriter#requestNewBufferBuilder(int targetChannel) throws IOException, InterruptedException {
		checkState(bufferBuilder == null || bufferBuilder.isFinished());

		BufferBuilder builder = super.requestNewBufferBuilder(targetChannel);
		if (randomTriggered) {
			addBufferConsumer(builder.createBufferConsumer(), targetChannel);
		} else {
			try (BufferConsumer bufferConsumer = builder.createBufferConsumer()) {
				for (int channel = 0; channel < numberOfChannels; channel++) {
					addBufferConsumer(bufferConsumer.copy(), channel);
				}
			}
		}

		bufferBuilder = builder;
		return builder;
	}
```
?